File: app/src/main/java/tabletop/velocic/com/worldforgerpgtools/NumGeneratorResultsFragment.java
package tabletop.velocic.com.worldforgerpgtools;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.NumberPicker;

public class NumGeneratorResultsFragment extends android.support.v4.app.DialogFragment
{
    private NumberPicker numberPicker;
    private String generatorPath;

    public static final String ARG_GENERATOR_PATH = "generator_path";
    public static final String EXTRA_NUM_GENERATOR_RESULTS = "tabletop.velocic.com.worldforgerpgtools.num_generator_results";
    public static final String EXTRA_GENERATOR_PATH = "tabletop.velocic.com.worldforgerpgtools.generator_path";

    public static NumGeneratorResultsFragment newInstance(String generatorPath)
    {
        Bundle args = new Bundle();
        args.putString(ARG_GENERATOR_PATH, generatorPath);

        NumGeneratorResultsFragment fragment = new NumGeneratorResultsFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState)
    {
        View view = LayoutInflater.from(getActivity())
            .inflate(R.layout.dialog_num_generator_results, null);

        numberPicker = view.findViewById(R.id.picker_num_generator_results);
        numberPicker.setMinValue(1);
        numberPicker.setMaxValue(999);

        Bundle fragmentArgs = getArguments();

        if (fragmentArgs != null) {
            generatorPath = fragmentArgs.getString(ARG_GENERATOR_PATH);
        }

        return new AlertDialog.Builder(getActivity())
            .setView(view)
            .setTitle(R.string.num_generator_results_dialog_title)
            .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener()
            {
                @Override
                public void onClick(DialogInterface dialogInterface, int i)
                {
                    sendResult(Activity.RESULT_OK, numberPicker.getValue());
                }
            })
            .setNegativeButton(android.R.string.cancel, new DialogInterface.OnClickListener()
            {
                @Override
                public void onClick(DialogInterface dialogInterface, int i)
                {
                    sendResult(Activity.RESULT_CANCELED, 0);
                }
            })
            .create();
    }

    private void sendResult(int resultCode, int numResultsToGenerate)
    {
        if (getTargetFragment() == null) {
            return;
        }

        Intent intent = new Intent();
        intent.putExtra(EXTRA_NUM_GENERATOR_RESULTS, numResultsToGenerate);
        intent.putExtra(EXTRA_GENERATOR_PATH, generatorPath);

        getTargetFragment().onActivityResult(getTargetRequestCode(), resultCode, intent);
    }
}

File: app/src/main/java/tabletop/velocic/com/worldforgerpgtools/NumGeneratorResultsFragment.kt
package tabletop.velocic.com.worldforgerpgtools

import android.app.Activity
import android.app.AlertDialog
import android.app.Dialog
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.dialog_num_generator_results.*

class NumGeneratorResultsFragment : android.support.v4.app.DialogFragment()
{
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
        inflater.inflate(R.layout.dialog_num_generator_results, container, false)

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        picker_num_generator_results.minValue = 1
        picker_num_generator_results.maxValue = 999

        val generatorPath = arguments?.getString(ARG_GENERATOR_PATH)
            ?: throw IllegalStateException("NumGeneratorResultsFragment requires a target Generator" +
                " path to invoke after a result has been selected, but none was provided.")

        val sendResult = { resultCode: Int, numResultsToGenerate: Int ->
            val intent = Intent().apply {
                putExtra(EXTRA_NUM_GENERATOR_RESULTS, numResultsToGenerate)
                putExtra(EXTRA_GENERATOR_PATH, generatorPath)
            }

            targetFragment?.onActivityResult(targetRequestCode, resultCode, intent)
                ?: throw IllegalStateException("Attempted to send NumGeneratorResultsFragment dialog" +
                    " result to the calling Fragment, however that fragment was somehow null.")
        }

        return AlertDialog.Builder(activity)
            .setView(view)
            .setTitle(R.string.num_generator_results_dialog_title)
            .setPositiveButton(android.R.string.ok) {_, _ -> sendResult(Activity.RESULT_OK, picker_num_generator_results.value)}
            .setNegativeButton(android.R.string.cancel) {_, _ -> sendResult(Activity.RESULT_CANCELED, 0)}
            .create()
    }

    companion object {
        const val ARG_GENERATOR_PATH = "generator_path"
        const val EXTRA_NUM_GENERATOR_RESULTS = "tabletop.velocic.com.worldforgerpgtools.num_generator_results"
        const val EXTRA_GENERATOR_PATH = "tabletop.velocic.com.worldforgerpgtools.generator_path"

        fun newInstance(generatorPath: String) : NumGeneratorResultsFragment
        {
            val fragment = NumGeneratorResultsFragment()

            fragment.arguments = Bundle().apply {
                putString(ARG_GENERATOR_PATH, generatorPath)
            }

            return fragment
        }
    }
}